core/unit_test/TestReduce.hpp

```diff
@@ -625,4 +625,63 @@ TEST(TEST_CATEGORY, int_combined_reduce_mixed) {
 }
 #endif
 #endif
+
+#if defined(NDEBUG)
+// the following test was made for:
+// https://github.com/kokkos/kokkos/issues/6517
+
+template <class T>
+struct EnableTestForReductionWithLargeIterationCount : std::false_type {};
+
+#ifdef KOKKOS_ENABLE_CUDA
+template <>
+struct EnableTestForReductionWithLargeIterationCount<Kokkos::Cuda>
+    : std::true_type {};
+#endif
+
+#ifdef KOKKOS_ENABLE_HIP
+template <>
+struct EnableTestForReductionWithLargeIterationCount<Kokkos::HIP>
+    : std::true_type {};
+#endif
+
+#ifdef KOKKOS_ENABLE_SYCL
+template <>
+struct EnableTestForReductionWithLargeIterationCount<Kokkos::Experimental::SYCL>
+    : std::true_type {};
+#endif
+
+#ifdef KOKKOS_ENABLE_OPENACC
+template <>
+struct EnableTestForReductionWithLargeIterationCount<
+    Kokkos::Experimental::OpenACC> : std::true_type {};
+#endif
+
+#ifdef KOKKOS_ENABLE_OPENMPTARGET
+template <>
+struct EnableTestForReductionWithLargeIterationCount<
+    Kokkos::Experimental::OpenMPTarget> : std::true_type {};
+#endif
+
+struct FunctorReductionWithLargeIterationCount {
+  KOKKOS_FUNCTION void operator()(const int64_t /*i*/, double& update) const {
+    update += 1.0;
+  }
+};
+
+TEST(TEST_CATEGORY, reduction_with_large_iteration_count) {
+  if constexpr (!EnableTestForReductionWithLargeIterationCount<
+                    TEST_EXECSPACE>::value) {
+    GTEST_SKIP() << "Disabling for host backends";
+  }
+
+  const int64_t N = pow(2LL, 39LL) - pow(2LL, 8LL) + 1;
+  Kokkos::RangePolicy<TEST_EXECSPACE, Kokkos::IndexType<int64_t>> p(0, N);
+  double nu = 0;
+  EXPECT_NO_THROW(Kokkos::parallel_reduce(
+      "sample reduction", p, FunctorReductionWithLargeIterationCount(), nu));
+  ASSERT_DOUBLE_EQ(nu, double(N));
```

Comment:
Does strict comparison actually fail?

I would have gone for reducing into a 64-bit integer that way we don't have to go through that macro and remember how many bits go to the mantissa of a IEEE double to be sure that we stay in the range of integers that can be represented exactly.

Reply:
I took the test from the issue that was posted without major changes

Reply:
I think it's unlikely to fail, but it doesn't hurt either. double is 11 bits for exponent. You take away 1 bit for sign too, that means up to 52bits of integral types should be fine I think?